js/__internal/viz/chart_components/m_advanced_chart.ts

```diff
@@ -0,0 +1,863 @@
+import { noop as _noop } from '@js/core/utils/common';
+import { extend as _extend } from '@js/core/utils/extend';
+import { each as _each, reverseEach as _reverseEach } from '@js/core/utils/iterator';
+import { isDefined as _isDefined, type } from '@js/core/utils/type';
+import { Axis } from '@js/viz/axes/base_axis';
+import { SeriesFamily } from '@js/viz/core/series_family';
+import {
+  convertVisualRangeObject, map as _map,
+  mergeMarginOptions, rangesAreEqual, setCanvasValues, unique,
+} from '@js/viz/core/utils';
+import rangeDataCalculator from '@js/viz/series/helpers/range_data_calculator';
+import { Range } from '@js/viz/translators/range';
+
+import { BaseChart } from './m_base_chart';
+
+const { isArray } = Array;
+
+const DEFAULT_AXIS_NAME = 'defaultAxisName';
+const FONT = 'font';
+const COMMON_AXIS_SETTINGS = 'commonAxisSettings';
+const DEFAULT_PANE_NAME = 'default';
+const VISUAL_RANGE = 'VISUAL_RANGE';
+
+function prepareAxis(axisOptions) {
+  return isArray(axisOptions) ? axisOptions.length === 0 ? [{}] : axisOptions : [axisOptions];
+}
+
+function processBubbleMargin(opt, bubbleSize) {
+  if (opt.processBubbleSize) {
+    opt.size = bubbleSize;
+  }
+  return opt;
+}
+
+function estimateBubbleSize(size, panesCount, maxSize, rotated) {
+  const width = rotated ? size.width / panesCount : size.width;
+  const height = rotated ? size.height : size.height / panesCount;
+
+  return Math.min(width, height) * maxSize;
+}
+
+function setAxisVisualRangeByOption(arg, axis, isDirectOption, index?) {
+  let options;
+  let visualRange;
+
+  if (isDirectOption) {
+    visualRange = arg.value;
+    options = {
+      skipEventRising: true,
+    };
+
+    const wrappedVisualRange = wrapVisualRange(arg.fullName, visualRange);
+    if (wrappedVisualRange) {
+      options = { allowPartialUpdate: true };
+      visualRange = wrappedVisualRange;
+    }
+  } else {
+    visualRange = (_isDefined(index) ? arg.value[index] : arg.value).visualRange;
+  }
+  axis.visualRange(visualRange, options);
+}
+
+function getAxisTypes(groupsData, axis, isArgumentAxes) {
+  if (isArgumentAxes) {
+    return { argumentAxisType: groupsData.argumentAxisType, argumentType: groupsData.argumentType };
+  }
+  const { valueAxisType, valueType } = groupsData.groups.filter((g) => g.valueAxis === axis)[0];
+  return { valueAxisType, valueType };
+}
+
+function wrapVisualRange(fullName, value) {
+  const pathElements = fullName.split('.');
+  const destElem = pathElements[pathElements.length - 1];
```

Comment:
Let's not shorten words
```suggestion
  const destinationElement = pathElements[pathElements.length - 1];
```